lib/rubocop/cop/style/fetch_env_var.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop suggests `ENV.fetch` for the replacement of `ENV[]`.
+      # `ENV[]` silently fails and returns `nil` when the environment variable is unset,
+      # which may cause unexpected behaviors when the developer forgets to set it.
+      # On the other hand, `ENV.fetch` raises KeyError or returns the explicitly
+      # specified default value.
+      #
+      # @example
+      #   # bad
+      #   ENV['X']
+      #   ENV['X'] || z
+      #   x = ENV['X']
+      #
+      #   # good
+      #   ENV.fetch('X')
+      #   ENV.fetch('X', nil) || z
+      #   x = ENV.fetch('X')
+      #
+      #   # also good
+      #   !ENV['X']
+      #   ENV['X'].some_method # (e.g. `.nil?`)
+      #
+      class FetchEnvVar < Base
+        extend AutoCorrector
+
+        MSG = 'Use `ENV.fetch(%<key>s)` or `ENV.fetch(%<key>s, nil)` instead of `ENV[%<key>s]`.'
+
+        # @!method env_with_bracket?(node)
+        def_node_matcher :env_with_bracket?, <<~PATTERN
+          (send (const nil? :ENV) :[] $_)
+        PATTERN
+
+        def on_send(node)
+          env_with_bracket?(node) do |expression|
+            break if allowed_var?(expression)
+            break unless offensive?(node)
+
+            add_offense(node, message: format(MSG, key: expression.source)) do |corrector|
+              corrector.replace(node, "ENV.fetch(#{expression.source}, nil)")
+            end
+          end
+        end
+
+        private
+
+        def allowed_var?(expression)
+          expression.str_type? && cop_config['AllowedVars'].include?(expression.value)
+        end
+
+        # rubocop:disable Metrics/CyclomaticComplexity
+        def offensive?(node)
+          only_node_of_expression?(node) ||
```

Comment:
I think it'd be a good idea to add some explanatory comment here, as it's not very clear why all those checks are needed.

Reply:
Great point. I will write some comment for that.

BTW, this cop ignore some `ENV[]` in the below cases;
- when an `ENV[]` is no more than a flag for a condition
- when an `ENV[]` receives some message (This may be controversial though)

This feature is documented in the `also good` section of the documentation.
I’m going to mention about above as well, but would like to hear if this idea sounds good for others.

Reply:
I added some comments and made the logic simpler.
https://github.com/rubocop/rubocop/pull/10502/commits/17a52efd5773c078698ed0d4bfa6dd4aad255f34